ddtrace/contrib/internal/google_adk/patch.py

```diff
@@ -0,0 +1,256 @@
+    instance = instance or args[0]
+
+    with integration.trace(
+        pin,
+        "%s.%s" % (instance.__class__.__name__, wrapped.__name__),
+        provider=provider_name,
+        model=model_name,
+        kind="tool",
+        submit_to_llmobs=True,
+    ) as span:
+        result = None
+        try:
+            result = await wrapped(*args, **kwargs)
+            return result
+        except Exception as e:
+            span.set_exc_info(type(e), e, e.__traceback__)
+            raise
+        finally:
+            span.set_tag_str("component", "google-adk")
+            kwargs["tool"] = args[0] if args else kwargs.get("tool")
+            kwargs["tool_args"] = args[1] if len(args) > 1 else kwargs.get("args")
+            kwargs["tool_call_id"] = tool_context.function_call_id
+            integration.llmobs_set_tags(
+                span,
+                args=args,
+                kwargs=kwargs,
+                response=result,
+            )
+
+
+@with_traced_module
+async def _traced_functions_call_tool_live(adk, pin, wrapped, instance, args, kwargs):
+    integration = adk._datadog_integration
+    tool_context = kwargs.get("tool_context", {})
+    agent = tool_context._invocation_context.agent
+    provider_name, model_name = extract_provider_and_model_name(agent=agent)
+
+    with integration.trace(
+        pin,
+        "%s.%s" % (instance.__class__.__name__, wrapped.__name__),
+        provider=provider_name,
+        model=model_name,
+        kind="tool",
+        submit_to_llmobs=True,
+    ) as span:
+        result = None
+        try:
+            agen = wrapped(*args, **kwargs)
+            async for item in agen:
+                yield item
+        except Exception as e:
+            span.set_exc_info(type(e), e, e.__traceback__)
+            raise
+        finally:
+            span.set_tag_str("component", "google-adk")
+            kwargs["tool"] = args[0] if args else kwargs.get("tool")
+            kwargs["tool_args"] = args[1] if len(args) > 1 else kwargs.get("args")
+            kwargs["tool_call_id"] = tool_context.function_call_id
+            integration.llmobs_set_tags(
+                span,
+                args=args,
+                kwargs=kwargs,
+                response=result,
+            )
+
+
+@with_traced_module
+def _traced_code_executor_execute_code(adk, pin, wrapped, instance, args, kwargs):
+    """Trace the execution of code by the agent (sync)."""
+    integration = adk._datadog_integration
+    invocation_context = kwargs.get("invocation_context", args[0] if args else None)
+    provider_name, model_name = extract_provider_and_model_name(agent=getattr(invocation_context, "agent", None))
+
+    # Signature: execute_code(self, invocation_context, code_execution_input)
+    code_input = kwargs.get("code_execution_input", args[1] if len(args) >= 2 and args[1] else None)
+    with integration.trace(
+        pin,
```

Comment:
🤔 required here? could be a blocker to removing `Pin` I guess ?

Reply:
technically required for now, but yeah in conjunction with your comment below on using the core api in another PR, we should be able to remove it